### PR TITLE
#393 test if calendar is not hidden in update action

### PR DESCRIPTION
**Pull request by asmaNode**
**DEMO: (for view calendar)**
https://user-images.githubusercontent.com/43316837/107039354-19d88700-67be-11eb-9d25-ad019ba9f55a.mp4

**DEMO :(for planning)**

https://user-images.githubusercontent.com/43316837/107700559-efa03100-6cb7-11eb-9b54-0855db824d5e.mp4

### CODE REVIEW

File: src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.js [hunk 1]
```diff
@@ -366,10 +366,12 @@ const _ = require('lodash');
...
-
-      calendarPromise.then(function(cal) {
-        cal.fullCalendar('removeEventSource', $scope.eventSourcesMap[calId]);
-        cal.fullCalendar('addEventSource', $scope.eventSourcesMap[calId]);
+      //Test if calendar is not hidden before
+      calendarVisibilityService.isHidden(calendar).then(function(calIsHidden) {
+        calendarPromise.then(function(cal) {
+          cal.fullCalendar('removeEventSource', $scope.eventSourcesMap[calId]);
+          !calIsHidden ? cal.fullCalendar('addEventSource', $scope.eventSourcesMap[calId]) : null;
+        });
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.js:374]:
```suggestion
      // remove and add the event source to redraw it
      calendarPromise.then((cal) => {
        cal.fullCalendar('removeEventSource', $scope.eventSourcesMap[calId]);
        // check if the calendar is hidden, if yes skip the redraw
        calendarVisibilityService.isHidden(calendar).then((isHidden) => {
          if (!isHidden) cal.fullCalendar('addEventSource', $scope.eventSourcesMap[calId]);
        })
      });
```

File: src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.spec.js [hunk 2]
```diff
@@ -429,7 +433,9 @@ describe('The calendarViewController', function() {
       expect(this.scope.calendars).to.shallowDeepEqual([this.calendars[0], { uniqueId: updatedCalendar.uniqueId, color: updatedCalendar.color }]);
       expect(this.scope.eventSourcesMap[updatedCalendar.getUniqueId()].backgroundColor).to.equal(updatedCalendar.color);
       expect(fullCalendarSpy).to.have.been.calledWith('removeEventSource', sinon.match.has('uniqueId', updatedCalendar.uniqueId));
-      expect(fullCalendarSpy).to.have.been.calledWith('addEventSource', sinon.match.has('uniqueId', updatedCalendar.uniqueId));
+      setTimeout(function() {
+        expect(fullCalendarSpy).to.have.been.calledWith('addEventSource', sinon.match.has('uniqueId', updatedCalendar.uniqueId));
+      }, 100);
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.spec.js:438]: Why use `setTimeout` here? You can use `$rootScope.$digest()` in the tests that have promises.

File: src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.spec.js [hunk 3]
```diff
@@ -417,6 +417,10 @@ describe('The calendarViewController', function() {
         }
       };
 
+      this.calendarVisibilityServiceMock.isHidden = sinon.stub();
+      this.calendarVisibilityServiceMock.isHidden
+        .returns(this.$q.when(false));
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.spec.js:422]:
```suggestion
      this.calendarVisibilityServiceMock.isHidden = sinon.stub().returns(this.$q.when(false));
```

File: src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.spec.js (hunk 2 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/calendar/calendar-view/calendar-view.controller.spec.js:438]: without a timeout fullCalendarSpy is always called with removeEventSource , i will try your solution

File: src/linagora.esn.calendar/app/planning/calendar-planning.controller.js [hunk 4]
```diff
@@ -90,7 +93,7 @@ const _ = require('lodash');
       });
     }
 
-    function updateCalendar(event, calendar) {
+    function updateCalendar(calendar) {
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/planning/calendar-planning.controller.js:96]: this change is wrong.
it is true that the `event` is not used here but it is needed to be the first argument and the calendar as the second argument.

if we remove it the function will use the event `CAL_EVENTS.CALENDARS.UPDATE` as a calendar here:
```js
if (calendar.getUniqueId() === cal.getUniqueId()) {
```

as if doing
```js
if (CAL_EVENTS.CALENDARS.UPDATE.getUniqueId() === cal.getUniqueId()) {
```

File: src/linagora.esn.calendar/app/planning/calendar-planning.controller.js [hunk 5]
```diff
@@ -53,7 +53,10 @@ const _ = require('lodash');
 
             eventSourcesMap[calId] = buildEventSourceForCalendar(calendar);
             calendarPromise.then(function(cal) {
-              cal.fullCalendar('addEventSource', eventSourcesMap[calId]);
+              calendarVisibilityService.isHidden(calendar).then(isHidden => {
+                if (!isHidden) cal.fullCalendar('addEventSource', eventSourcesMap[calId]);
+              });
+              cal.fullCalendar('removeEventSource', eventSourcesMap[calId]);
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/planning/calendar-planning.controller.js:59]: I suggest calling the `removeEventSource` first ( even tho it's going to be called first in your current code cause javascript won't wait for the `calendarVisibilityService`  call ). for me it looks more clear if we reflect it in the code 

```suggestion
              cal.fullCalendar('removeEventSource', eventSourcesMap[calId]); // remove it first
              calendarVisibilityService.isHidden(calendar).then(isHidden => {
                // if it's not hidden add it again to redraw
                if (!isHidden) cal.fullCalendar('addEventSource', eventSourcesMap[calId]);
              });

```

File: src/linagora.esn.calendar/app/planning/calendar-planning.controller.js [hunk 6]
```diff
@@ -53,7 +53,10 @@ const _ = require('lodash');
 
             eventSourcesMap[calId] = buildEventSourceForCalendar(calendar);
             calendarPromise.then(function(cal) {
-              cal.fullCalendar('addEventSource', eventSourcesMap[calId]);
+              calendarVisibilityService.isHidden(calendar).then(isHidden => {
```
Comment (rezk2ll) [inline src/linagora.esn.calendar/app/planning/calendar-planning.controller.js:56]: the test for the changes in this file are missing

File: src/linagora.esn.calendar/app/planning/calendar-planning.controller.js (hunk 4 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/planning/calendar-planning.controller.js:96]: ok , we must first check this part of code , this function i think it is never called

**Review by rezk2ll — APPROVED** [review]

**Review by huy-ta — APPROVED** [review]

**Pull request merged**